plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -598,25 +604,76 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String key = keyFromPath(path);
+        if (recursive) {
+            boolean deleted = deletePrefix(path);
```

Comment:
I think the logic around `deletePrefix(path)` is sketchy here, because `deletePrefix` will return `false` if any objects failed to delete. I guess before we also would have returned `true` if any of the individual deletes failed, so nothing ultimately has changed- but below we decide whether to call `deleteObject(key)` based on whether _everything_ was deleted as expected. That's maybe justifiable, but then we should return `false` from this method if the intention was to leave the directory in tact. So here are the options that seem reasonable to me:
- Remove the `if (!deleted)` check guarding `deleteObject(key)`, call it unconditionally regardless of whether there were sub-objects deleted. Either continue always returning `true` from the method, or return whether any of the deletes deleted at least one value.
- Have some return value signal for failures having occurred in `deletePrefix` and avoid calling `deleteObject(key)` or `deleteObject(key + DIRECTORY_SUFFIX)` when failures occurred, return false from the method. Alternatively, proceed with the attempt to `deleteObject(key)` and `deleteObject(key + DIRECTORY_SUFFIX)` regardless of failures and return `true` if at least one of those three deletes succeeded.
-

Reply:
I've refactored the logic to take indeed note of this change. Sorry for overseeing it while removing the retries.

Reply:
ultimately, `io.trino.filesystem.TrinoFileSystem#deleteDirectory` doesn't return a `boolean` -- it should throw if anything left behind

for `org.apache.hadoop.fs.FileSystem#delete(org.apache.hadoop.fs.Path, boolean)` we should return true if and only if everything deleted, i.e. return false  if anything left behind